Userland/Libraries/LibSQL/Database.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2021, Jan de Visser <jan@de-visser.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Format.h>
+#include <AK/RefPtr.h>
+#include <AK/String.h>
+
+#include <LibSQL/BTree.h>
+#include <LibSQL/Database.h>
+#include <LibSQL/Heap.h>
+#include <LibSQL/Key.h>
+#include <LibSQL/Meta.h>
+#include <LibSQL/Tuple.h>
+
+namespace SQL {
+
+Database::Database(String name)
+{
+    m_heap = Heap::construct(move(name));
+    m_tables = BTree::construct(*m_heap, TableDef::index_def(), m_heap->tables_root());
+    m_tables->on_new_root = [&]() {
+        m_heap->set_tables_root(m_tables->root());
+    };
+    m_table_columns = BTree::construct(*m_heap, ColumnDef::index_def(), m_heap->table_columns_root());
+    m_table_columns->on_new_root = [&]() {
+        m_heap->set_table_columns_root(m_table_columns->root());
+    };
+}
+
+void Database::add_table(TableDef& table)
+{
+    m_tables->insert(table.key());
+    for (auto& column : table.columns()) {
+        m_table_columns->insert(column.key());
+    }
+}
+
+Optional<Key> Database::get_table_key(String table_name)
+{
+    auto n = move(table_name);
+    auto key = TableDef::make_key();
+    key["table_name"] = n;
+    auto table_iterator = m_tables->find(key);
+    if (table_iterator.is_end() || (String)((*table_iterator)[0]) != n) {
+        warnln("Table {} not found", n);
+        return {};
+    }
+    return *table_iterator;
+}
+
+RefPtr<TableDef> Database::get_table(String name)
+{
+    auto n = move(name);
+    auto optional_iter = get_table_key(n);
```

Comment:
We move to `n` then copy `n` when we call `get_table_key`, can we just move `name` into the call?
```suggestion
    auto optional_iter = get_table_key(move(name));
```

Reply:
Same as above haha